core/kotlinx-coroutines-core/src/test_/TestCoroutineScope.kt

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2016-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.test
+
+import kotlinx.coroutines.*
+import kotlin.coroutines.ContinuationInterceptor
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * A scope which provides detailed control over the execution of coroutines for tests.
+ *
+ * @param context an optional context that must provide delegates [ExceptionCaptor] and [DelayController]
+ */
+class TestCoroutineScope(
```

Comment:
Since it's a recommended practice to have classes implement `CoroutineScope` to provide convenient access to coroutine builders, it would be nice if the same pattern could be used for tests. E.g.

```kotlin
class MyTest : TestCoroutineScope {
  override val context = TestCoroutineDispatcher() + SomeCustomElement()

  @Test fun thingsWork() = asyncTest {
    …
  }
}
```

The way this API is designed, it requires custom scopes to always be expressed as explicit properties. Is this required verbosity intentional?

Reply:
I'm not a huge fan of that style in tests, since `TestCoroutineDispatcher` is stateful and it wouldn't take much to leak state between tests if coded like that.

However, if you preferred to code it that way, it would just require a this to ensure you're dispatching to the builder you intend, as well as ensure that dispatcher.cleanupTestCoroutines gets called @After.

From a real-world test perspective, using something like a JUnit4 rule would probably be the way to go here, which would already move out of this inheritance model.

```
@Test fun things() = this.asyncTest { }
```

Reply:
Agreed a rule is probably better. Figured this was intended to prevent that pattern, but since that's a pretty common pattern in the official docs it might be helpful to explicitly call it out in the docs (e.g. "Implementing the scope interface is not recommended, for {reasons}, instead use this pattern.").